### ISSUE TITLE
Passing environment variables through ?

### ISSUE THREAD

**Issue by eddelbuettel** [PrairieLearn/PrairieLearn#4733]
We have an open PR in our nascent 'R question demos' repo where I support 'debug mode' by setting a toggle in `run.sh`, see https://github.com/eddelbuettel/pl-r-demos/pull/2

Now, Docker itself lets us set environment variables at the command-line and that is use by numerous containers to great effect.  But here we are in a Docker-in-Docker situation that I am less familiar with.  Is there a best practices standard for passing (some, all, any, none, ...) environment variable from the (outer) call to the (inner) container running inside the container?  It would be sweet to just turn 'verbose' or 'debug' mode on but I haven't really thought through if we'd unleash any undesirable monsters if we permitted that as a more general measure.  Does anybody here know?

(I figure I can also proxy it by just `touch`-ing a flag file inside the work directory and tickling a change a behavior is such a file is seen.  Possibly safer and likely simpler.)

**Comment by echuber2** [comment]
You can set them on the entrypoint in info.json, here: https://github.com/eddelbuettel/pl-r-demos/blob/feature/update_run_sh/questions/R_Demos/test_R_function/info.json

But I'm not sure I fully understand the use case you had in mind.

**Comment by echuber2** [comment]
e.g. if `/tmp/foo.sh` contains:

```
#!/bin/bash
echo "$ZOUNDS"
```

Then you can set the entrypoint like:

```
ZOUNDS="Spaceman Spiff" /tmp/foo.sh
```

**Comment by eddelbuettel** [comment]
I call `prairielearn/prairielearn` via `docker` and can give `-e foo=bar` all day long.  

Now `prairielearn/prairielearn` calls, in my course, `stat447/pl` but  does _not_ as far as I understand it pass environments variable on when it does the black magic of docker-in-docker.  I could be wrong, I didn't test.   Will `stat447/pl` see `${foo}` ?

**Comment by echuber2** [comment]
No, the autograder or workspace docker will not automatically inherit environment variables from the PrairieLearn server. However you can arrange for options to be passed through other means.

The workspace has an "args" option you can set, which will be passed to your entrypoint. You would have to design the entrypoint to parse the args and do something meaningful (for example, unpack and apply environment variables). For the external grader, there is no "args" option currently. Instead you may have to set special options on the `data` object that is handed to the grading job (this shows up as `/grade/data/data.json` in the grader). Or if it's a hardcoded toggle you want to set locally when debugging, you could set options in some `serverFilesCourse` that are always sourced, as these also show up in the grader.

**Comment by nwalters512** [comment]
I think it would be reasonable to add an `environment` option to `workspaceOptions` and `externalGradingOptions` in `infoQuestion.json` - @eddelbuettel would that work for you? Something like this:

```json
{
  "uuid": "...",
  "externalGradingOptions": {
    "image": "...",
    "environment": {
      "DEBUG": "on"
    }
  },
  "workspaceOptions": {
    "image": "...",
    "environment": {
      "DEBUG": "on"
    }
  }
}
```

**Comment by eddelbuettel** [comment]
"Probably" - it is something like that this that I had in mind.  Might it be of enough general interest?

**Comment by nwalters512** [comment]
I suspect so - probably not broadly useful for external grading, but for workspaces, I've certainly seen third-party Docker images that are configured with various environment variables, and it may be useful to be able to do that without having to build an image on top of them with `ENV` statements.

This suggestion is specifically an alternative to trying to "pass environment variables through", or inherit them from the shell you're running things in. That's a recipe for nondeterministic behavior, where you or some TA has set some environment variable that PrairieLearn doesn't have in production, and things break or misbehave in unexplainable ways. I'd much rather that environment variables be explicitly set, as in the above proposal.

**Comment by eddelbuettel** [comment]
To be plain: what you go in the second paragraph is not what I asked for.  But putting parameters into the 'outer' (or initial if you will)  call to the prarielearn container is both amply used and documented.  Given that we run docker-in-docker here it seems fair to do the see same for the inner call.  And putting it into the json appears to both be the right place, and eminently feasible.

**Comment by nwalters512** [comment]
I brought that up because PrairieLearn has no way of distinguishing where environment variables come from; one that was set with `docker run ... -e ...` looks the same to it as the rest of its environment variables. At any rate: seems like we're on the same page now and have a route forward if anyone wants to pick up this issue!

**Comment by nwalters512** [comment]
The PR implementing this has been merged, so it's available to test in the base image, but it hasn't been deployed yet. Keep an eye on Slack to see when it's ready in production!
